### PR TITLE
Signals: Port connect calls to use callable_mp

### PR DESCRIPTION
**Pull request by akien-mga**
Follow-up to #36393 (and fixes a couple issues missed before merging it).

Remove now unnecessary bindings of signal callbacks in the public API.

No regular expressions were harmed in the making of this commit.
(Nah, just kidding.)

----

This is the output of the wonderfully hacky bash script:
```bash
#!/bin/bash

rm -f tmp-files tmp-file-matches tmp-errors

rg -g'!thirdparty' -g'!*.sh' -l "connecte?d?_compat\(" > tmp-files

# Associative array, bash extension.
# Keys will be "$class::$method_str" and values are "$method_cpp"
declare -A METHOD_MAP

while IFS= read -r file; do
  echo -e "\n### $file\n"
  grep "connecte\?d\?_compat(" "$file" > tmp-file-matches

  while IFS= read -r line; do
    echo "---------"
    echo $line
    # What base connect method are we replacing?
    connect=$(echo $line | sed -n 's/^.*\(connecte\?d\?\)_compat.*$/\1/p')
    # Restrict to sane part to avoid choking on stuff like [].
    first_part=$(echo $line | sed -n "s/^.*\(${connect}_compat(\"[^\"]*\", this, \"[^\"]*\"\).*/\1/p")
    echo -e "first part: $first_part"
    if [ -z "$first_part" ]; then
      # Likely not using 'this', skip.
      continue
    fi
    # Get method String name from the bindings.
    method_str=$(echo $first_part | sed "s/.*, this, \"\([^\"]*\)\".*/\1/")
    echo -e "method str: $method_str"
    # Look up from the signal connection to figure out which class we're in.
    # We only handle connections to 'this'.
    class=$(sed "/$first_part/q" "$file" | sed -n 's/^\(\S* \)\?\*\?\(\S*\)::.*$/\2/p' | tail -n 1)
    echo -e "class: $class"
    # Check if we retrieved this already.
    key="$class::$method_str"
    method_cpp=
    if [ ${METHOD_MAP[$key]+_} ]; then
      # Use saved value.
      method_cpp=${METHOD_MAP[$key]}
    else
      # Figure out the C++ method from the (hopefully matching) bind_method call.
      method_cpp=$(sed -n "s/.*ClassDB::bind_method(.*\"${method_str}\".*, \&${class}::\(\S*\)[,)].*/\1/p" "$file")
      if [ ! -z "$method_cpp" ]; then
        # Save the value we found for later use, and delete binding.
        METHOD_MAP[$key]=$method_cpp
        sed "/ClassDB::bind_method(.*\"${method_str}\".*, \&${class}::${method_cpp}/d" -i "$file"
      else
        echo "## ERROR: Method not found ##" >> tmp-errors
        echo -e "$file\n$line\nfirst part: $first_part\nmethod str: $method_str\nclass: $class\n" >> tmp-errors
        continue
      fi
    fi
    echo -e "method cpp: $method_cpp"
    # Do the final replace for the _compat method.
    sed -z "/$first_part/s/${connect}_compat(\"\([^\"]*\)\", this, \"${method_str}\"/${connect}(\"\1\", callable_mp(this, \&${class}::${method_cpp})/" -i "$file"
    # Add (method_str -> method_cpp) conversion to lookup array, and remove binding.
  done < "tmp-file-matches"
done < "tmp-files"
```

There are still a few cases which I now need to review manually (namely signals connected to something else than `this`, and signals connected to a method bound in a parent class).

### CODE REVIEW

**Comment by akien-mga** [comment]
Fixed my script which missed a lot of bindings to remove. There might be false positives in the lot though.

**Comment by akien-mga** [comment]
I ported most uses of `connect_compat` that I could find. As mentioned in the last commit, some cases can't be converted as they use a private or protected method from a different classes, which `callable_mp` does not support.

For example:
```
editor/create_dialog.cpp
807:    help_bit->connect_compat("request_hide", this, "_closed");
```
`this` is a `CreateDialog` which inherits `_closed` from `WindowDialog`. But `callable_mp` doesn't allow using it, even if we cast with `help_bit->connect("request_hide", callable_mp((WindowDialog *)this, &WindowDialog::_closed)`.

So if we want to convert those, we need to either make `callable_mp` support calling protected methods from parent classes, or make them public.

Another quirk of the new system is shown with e.g.:
```
gn->connect("item_rect_changed", callable_mp((CanvasItem *)connections_layer, &CanvasItem::update));
```
To call a public method from a parent class, we're forced to cast the object.

Finally, while everything compiles fine, opening a new project triggers a segfault:
```
[1] /lib64/libc.so.6() [0x38cbc3caf0] (??:0)
[2] Node::is_inside_tree() const (/home/akien/Projects/godot/godot.git/./scene/main/node.h:290)
[3] Control::grab_focus() (/home/akien/Projects/godot/godot.git/scene/gui/control.cpp:2169)
[4] EditorSettingsDialog::_focus_current_search_box() (/home/akien/Projects/godot/godot.git/editor/settings_config_dialog.cpp:371)
[5] EditorSettingsDialog::_tabs_tab_changed(int) (/home/akien/Projects/godot/godot.git/editor/settings_config_dialog.cpp:358)
[6] void call_with_variant_args_helper<EditorSettingsDialog, int, 0ul>(EditorSettingsDialog*, void (EditorSettingsDialog::*)(int), Variant const**, Callable::CallError&, IndexSequence<0ul>) (/home/akien/Projects/godot/godot.git/./core/callable_method_pointer.h:139 (discriminator 4))
[7] void call_with_variant_args<EditorSettingsDialog, int>(EditorSettingsDialog*, void (EditorSettingsDialog::*)(int), Variant const**, int, Callable::CallError&) (/home/akien/Projects/godot/godot.git/./core/callable_method_pointer.h:161)
[8] CallableCustomMethodPointer<EditorSettingsDialog, int>::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/akien/Projects/godot/godot.git/./core/callable_method_pointer.h:177)
[9] Callable::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/akien/Projects/godot/godot.git/core/callable.cpp:54)
[10] Object::emit_signal(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot.git/core/object.cpp:1200)
[11] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:1256)
[12] TabContainer::add_child_notify(Node*) (/home/akien/Projects/godot/godot.git/scene/gui/tab_container.cpp:542 (discriminator 4))
[13] Node::_add_child_nocheck(Node*, StringName const&) (/home/akien/Projects/godot/godot.git/scene/main/node.cpp:1248)
[14] Node::add_child(Node*, bool) (/home/akien/Projects/godot/godot.git/scene/main/node.cpp:1260)
[15] EditorSettingsDialog::EditorSettingsDialog() (/home/akien/Projects/godot/godot.git/editor/settings_config_dialog.cpp:409)
[16] EditorNode::EditorNode() (/home/akien/Projects/godot/godot.git/editor/editor_node.cpp:6070)
[17] Main::start() (/home/akien/Projects/godot/godot.git/main/main.cpp:1742)
[18] /home/akien/Projects/godot/godot.git/bin/godot.x11.tools.64(main+0xe7) [0x17431b9] (/home/akien/Projects/godot/godot.git/platform/x11/godot_x11.cpp:55)
[19] /lib64/libc.so.6(__libc_start_main+0xeb) [0x38cbc26b0b] (??:0)
[20] /home/akien/Projects/godot/godot.git/bin/godot.x11.tools.64(_start+0x2a) [0x174302a] (/home/iurt/rpmbuild/BUILD/glibc-2.29/csu/../sysdeps/x86_64/start.S:122)
-- END OF BACKTRACE --
```

Over to you @reduz :)

You can checkout this branch locally with:
```
git checkout -b akien-mga-calling-all-stations master
git pull https://github.com/akien-mga/godot.git calling-all-stations
```

**Comment by kuruk-mm** [comment]
The next time, if you want someone to do repeated work let me know. I'm your human script haha. I want to help.

**Comment by kuruk-mm** [comment]
> I ported most uses of `connect_compat` that I could find. As mentioned in the last commit, some cases can't be converted as they use a private or protected method from a different classes, which `callable_mp` does not support.
> 
> For example:
> 
> ```
> editor/create_dialog.cpp
> 807:    help_bit->connect_compat("request_hide", this, "_closed");
> ```
> 
> `this` is a `CreateDialog` which inherits `_closed` from `WindowDialog`. But `callable_mp` doesn't allow using it, even if we cast with `help_bit->connect("request_hide", callable_mp((WindowDialog *)this, &WindowDialog::_closed)`.
> 
> So if we want to convert those, we need to either make `callable_mp` support calling protected methods from parent classes, or make them public.
> 
> Another quirk of the new system is shown with e.g.:
> 
> ```
> gn->connect("item_rect_changed", callable_mp((CanvasItem *)connections_layer, &CanvasItem::update));
> ```
> 
> To call a public method from a parent class, we're forced to cast the object.
> 
> Finally, while everything compiles fine, opening a new project triggers a segfault:
> 
> ```
> [1] /lib64/libc.so.6() [0x38cbc3caf0] (??:0)
> [2] Node::is_inside_tree() const (/home/akien/Projects/godot/godot.git/./scene/main/node.h:290)
> [3] Control::grab_focus() (/home/akien/Projects/godot/godot.git/scene/gui/control.cpp:2169)
> [4] EditorSettingsDialog::_focus_current_search_box() (/home/akien/Projects/godot/godot.git/editor/settings_config_dialog.cpp:371)
> [5] EditorSettingsDialog::_tabs_tab_changed(int) (/home/akien/Projects/godot/godot.git/editor/settings_config_dialog.cpp:358)
> [6] void call_with_variant_args_helper<EditorSettingsDialog, int, 0ul>(EditorSettingsDialog*, void (EditorSettingsDialog::*)(int), Variant const**, Callable::CallError&, IndexSequence<0ul>) (/home/akien/Projects/godot/godot.git/./core/callable_method_pointer.h:139 (discriminator 4))
> [7] void call_with_variant_args<EditorSettingsDialog, int>(EditorSettingsDialog*, void (EditorSettingsDialog::*)(int), Variant const**, int, Callable::CallError&) (/home/akien/Projects/godot/godot.git/./core/callable_method_pointer.h:161)
> [8] CallableCustomMethodPointer<EditorSettingsDialog, int>::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/akien/Projects/godot/godot.git/./core/callable_method_pointer.h:177)
> [9] Callable::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/akien/Projects/godot/godot.git/core/callable.cpp:54)
> [10] Object::emit_signal(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot.git/core/object.cpp:1200)
> [11] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:1256)
> [12] TabContainer::add_child_notify(Node*) (/home/akien/Projects/godot/godot.git/scene/gui/tab_container.cpp:542 (discriminator 4))
> [13] Node::_add_child_nocheck(Node*, StringName const&) (/home/akien/Projects/godot/godot.git/scene/main/node.cpp:1248)
> [14] Node::add_child(Node*, bool) (/home/akien/Projects/godot/godot.git/scene/main/node.cpp:1260)
> [15] EditorSettingsDialog::EditorSettingsDialog() (/home/akien/Projects/godot/godot.git/editor/settings_config_dialog.cpp:409)
> [16] EditorNode::EditorNode() (/home/akien/Projects/godot/godot.git/editor/editor_node.cpp:6070)
> [17] Main::start() (/home/akien/Projects/godot/godot.git/main/main.cpp:1742)
> [18] /home/akien/Projects/godot/godot.git/bin/godot.x11.tools.64(main+0xe7) [0x17431b9] (/home/akien/Projects/godot/godot.git/platform/x11/godot_x11.cpp:55)
> [19] /lib64/libc.so.6(__libc_start_main+0xeb) [0x38cbc26b0b] (??:0)
> [20] /home/akien/Projects/godot/godot.git/bin/godot.x11.tools.64(_start+0x2a) [0x174302a] (/home/iurt/rpmbuild/BUILD/glibc-2.29/csu/../sysdeps/x86_64/start.S:122)
> -- END OF BACKTRACE --
> ```
> 
> Over to you @reduz :)
> 
> You can checkout this branch locally with:
> 
> ```
> git checkout -b akien-mga-calling-all-stations master
> git pull https://github.com/akien-mga/godot.git calling-all-stations
> ```

I analyzed this. Comparing the behavior before and after the new signal system. I think this is a side effect of the new signal system.

This particle reason is because this is using 'search_box' before it is instanciate.

image

And now, with the new signal system, we have available the functions when it connect to the method. With the old signal system we have available when it execute "_bind_methods" (it is calling as postconstructor) to have it available.

So my proposals for this is:
1. Setting the connections on bottom of the constructor. 
2. Review every case and take care of this.
3. Create a method disable_signals() and enable_signals(), and disable on the constructor, and enable on the bottom of the constructor.

**Comment by akien-mga** [comment]
Thanks to @kuruk-mm this is no longer crashing :)

There are still issues to find and fix, but it can get some more broader testing if anyone is interested.

**Comment by kuruk-mm** [comment]
> are still issues to find and fix, but it can get some more broader testing if anyone is interes

I'm on it

**Comment by akien-mga** [comment]
I reviewed the list of bindings removed in the first commit that don't start with a `_` (potentially public methods) and fixed a few issues:
```
-	ClassDB::bind_method("start", &EditorDebuggerNode::start);
-	ClassDB::bind_method("stop", &EditorDebuggerNode::stop);
-	ClassDB::bind_method("request_remote_tree", &EditorDebuggerNode::request_remote_tree);
-	ClassDB::bind_method(D_METHOD("debug_skip_breakpoints"), &ScriptEditorDebugger::debug_skip_breakpoints);
-	ClassDB::bind_method(D_METHOD("debug_copy"), &ScriptEditorDebugger::debug_copy);
-	ClassDB::bind_method(D_METHOD("debug_next"), &ScriptEditorDebugger::debug_next);
-	ClassDB::bind_method(D_METHOD("debug_step"), &ScriptEditorDebugger::debug_step);
-	ClassDB::bind_method(D_METHOD("debug_break"), &ScriptEditorDebugger::debug_break);
-	ClassDB::bind_method(D_METHOD("debug_continue"), &ScriptEditorDebugger::debug_continue);
-	ClassDB::bind_method(D_METHOD("reload"), &EditorDirDialog::reload, DEFVAL(""));
-	ClassDB::bind_method("update_plugins", &EditorPluginSettings::update_plugins);
-	ClassDB::bind_method("update_tree", &GroupsEditor::update_tree);
-	ClassDB::bind_method(D_METHOD("show_groups"), &NodeDock::show_groups);
-	ClassDB::bind_method(D_METHOD("show_connections"), &NodeDock::show_connections);
-	ClassDB::bind_method("apply_shaders", &ShaderEditor::apply_shaders);
-	ClassDB::bind_method("save_external_data", &ShaderEditor::save_external_data);
-	ClassDB::bind_method(D_METHOD("update_transform_gizmo_view"), &SpatialEditorViewport::update_transform_gizmo_view);
-	ClassDB::bind_method("reset", &RenameDialog::reset);
-	ClassDB::bind_method(D_METHOD("set_hsv_mode", "mode"), &ColorPicker::set_hsv_mode);
-	ClassDB::bind_method(D_METHOD("set_raw_mode", "mode"), &ColorPicker::set_raw_mode);
-	ClassDB::bind_method(D_METHOD("queue_sort"), &Container::queue_sort);
-	ClassDB::bind_method(D_METHOD("deselect_items"), &FileDialog::deselect_items);
-	ClassDB::bind_method(D_METHOD("menu_option", "option"), &LineEdit::menu_option);
-	ClassDB::bind_method(D_METHOD("menu_option", "option"), &TextEdit::menu_option);
```

Also fixed start errors which pointed to other wrongly removed bindings:
```
ERROR: Invalid setter 'FileDialog::set_show_hidden_files' for property 'show_hidden_files'.
   at: add_property (core/class_db.cpp:945)
ERROR: Invalid setter 'ColorPicker::set_hsv_mode' for property 'hsv_mode'.
   at: add_property (core/class_db.cpp:945)
ERROR: Invalid setter 'ColorPicker::set_raw_mode' for property 'raw_mode'.
   at: add_property (core/class_db.cpp:945)
ERROR: Invalid setter 'EditorFileDialog::set_display_mode' for property 'display_mode'.
   at: add_property (core/class_db.cpp:945)
ERROR: Invalid setter 'EditorFileDialog::set_show_hidden_files' for property 'show_hidden_files'.
   at: add_property (core/class_db.cpp:945)
```

**Comment by akien-mga** [comment]
There are still those new errors raised when opening a project:
```
ERROR: Condition "is_custom()" is true. returned: StringName()
   at: get_method (core/callable.cpp:76)
ERROR: Condition "is_custom()" is true. returned: StringName()
   at: get_method (core/callable.cpp:76)
```

**Comment by akien-mga** [comment]
```
#4  0x0000000002e419ef in ConnectionsDock::update_tree (this=0x917ad30) at editor/connections_dialog.cpp:1018
```

Using `gdb` I could assess that those two errors were raised for the signals `visibility_changed` and `script_changed` connected in `SceneTreeEditor`.

Commenting those out then led me to find more errors when adding nodes to the scene tree. The diff below seems to silence all of them, though it obviously breaks the editor.

```diff
diff --git a/core/callable.cpp b/core/callable.cpp
index 34b79cea10..a25830bf17 100644
--- a/core/callable.cpp
+++ b/core/callable.cpp
@@ -73,7 +73,8 @@ ObjectID Callable::get_object_id() const {
 	}
 }
 StringName Callable::get_method() const {
-	ERR_FAIL_COND_V(is_custom(), StringName());
+	ERR_FAIL_COND_V_MSG(is_custom(), StringName(),
+			vformat("Can't get method on CallableCustom \"%s\".", operator String()));
 	return method;
 }
 uint32_t Callable::hash() const {
diff --git a/editor/editor_data.cpp b/editor/editor_data.cpp
index 26d132665c..01a4cca5ed 100644
--- a/editor/editor_data.cpp
+++ b/editor/editor_data.cpp
@@ -1024,7 +1024,9 @@ void EditorSelection::add_node(Node *p_node) {
 	}
 	selection[p_node] = meta;
 
+	/*
 	p_node->connect("tree_exiting", callable_mp(this, &EditorSelection::_node_removed), varray(p_node), CONNECT_ONESHOT);
+	*/
 
 	//emit_signal("selection_changed");
 }
@@ -1042,7 +1044,9 @@ void EditorSelection::remove_node(Node *p_node) {
 	if (meta)
 		memdelete(meta);
 	selection.erase(p_node);
+	/*
 	p_node->disconnect("tree_exiting", callable_mp(this, &EditorSelection::_node_removed));
+	*/
 	//emit_signal("selection_changed");
 }
 bool EditorSelection::is_selected(Node *p_node) const {
diff --git a/editor/scene_tree_editor.cpp b/editor/scene_tree_editor.cpp
index e4e642e368..fb5d9cecd9 100644
--- a/editor/scene_tree_editor.cpp
+++ b/editor/scene_tree_editor.cpp
@@ -323,8 +323,10 @@ bool SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 
 	if (can_open_instance && undo_redo) { //Show buttons only when necessary(SceneTreeDock) to avoid crashes
 
+		/*
 		if (!p_node->is_connected("script_changed", callable_mp(this, &SceneTreeEditor::_node_script_changed)))
 			p_node->connect("script_changed", callable_mp(this, &SceneTreeEditor::_node_script_changed), varray(p_node));
+		*/
 
 		Ref<Script> script = p_node->get_script();
 		if (!script.is_null()) {
@@ -350,8 +352,10 @@ bool SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 			else
 				item->add_button(0, get_icon("GuiVisibilityHidden", "EditorIcons"), BUTTON_VISIBILITY, false, TTR("Toggle Visibility"));
 
+			/*
 			if (!p_node->is_connected("visibility_changed", callable_mp(this, &SceneTreeEditor::_node_visibility_changed)))
 				p_node->connect("visibility_changed", callable_mp(this, &SceneTreeEditor::_node_visibility_changed), varray(p_node));
+			*/
 
 			_update_visibility_color(p_node, item);
 		} else if (p_node->is_class("Spatial")) {
@@ -370,8 +374,10 @@ bool SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 			else
 				item->add_button(0, get_icon("GuiVisibilityHidden", "EditorIcons"), BUTTON_VISIBILITY, false, TTR("Toggle Visibility"));
 
+			/*
 			if (!p_node->is_connected("visibility_changed", callable_mp(this, &SceneTreeEditor::_node_visibility_changed)))
 				p_node->connect("visibility_changed", callable_mp(this, &SceneTreeEditor::_node_visibility_changed), varray(p_node));
+			*/
 
 			_update_visibility_color(p_node, item);
 		} else if (p_node->is_class("AnimationPlayer")) {
diff --git a/scene/gui/control.cpp b/scene/gui/control.cpp
index 1a231e368b..de6737ec4f 100644
--- a/scene/gui/control.cpp
+++ b/scene/gui/control.cpp
@@ -542,7 +542,9 @@ void Control::_notification(int p_notification) {
 
 				if (data.parent_canvas_item) {
 
+					/*
 					data.parent_canvas_item->connect("item_rect_changed", callable_mp(this, &Control::_size_changed));
+					*/
 				} else {
 					//connect viewport
 					get_viewport()->connect("size_changed", callable_mp(this, &Control::_size_changed));
@@ -561,7 +563,9 @@ void Control::_notification(int p_notification) {
 
 			if (data.parent_canvas_item) {
 
+				/*
 				data.parent_canvas_item->disconnect("item_rect_changed", callable_mp(this, &Control::_size_changed));
+				*/
 				data.parent_canvas_item = NULL;
 			} else if (!is_set_as_toplevel()) {
 				//disconnect viewport
```

**Comment by akien-mga** [comment]
Alright with help from @reduz I could fix those errors: https://github.com/godotengine/godot/pull/36426#issuecomment-592210628

From some quick testing everything seems to behave OK. I'm pretty sure there might still be situations where a removed binding would still be needed by MessageQueue or UndoRedo, but we can assess those as errors pop up.

**Comment by kuruk-mm** [comment]
I made a simple 2D game and works fine.
And I tested master and your PR with (https://github.com/qarmin/The-worst-Godot-test-project) and I get similar behavior without extra errors from the master.

So, I think this is working fine

**Comment by akien-mga** [comment]
Thanks a lot for the help @kuruk-mm!

File: editor/groups_editor.cpp [hunk 1]
```diff
@@ -668,12 +659,6 @@ void GroupsEditor::_show_group_dialog() {
 }
 
 void GroupsEditor::_bind_methods() {
-
-	ClassDB::bind_method("_add_group", &GroupsEditor::_add_group);
-	ClassDB::bind_method("_remove_group", &GroupsEditor::_remove_group);
-	ClassDB::bind_method("update_tree", &GroupsEditor::update_tree);
```
Comment (nathanwfranke) [inline editor/groups_editor.cpp]: Why were these removed? This is now causing an error when I add/remove an object from a group

```
ERROR: Error calling method from signal 'update_tree': 'GroupsEditor::update_tree': Method not found.
   at: _process_operation_list (core/undo_redo.cpp:296)
```

Error goes away if I add `ClassDB::bind_method("update_tree", &GroupsEditor::update_tree);`

Comment (akien-mga) [reply editor/groups_editor.cpp]: I thought I had undone that change after https://github.com/godotengine/godot/pull/36426#issuecomment-592209600 but it seems I missed it.

Comment (akien-mga) [reply editor/groups_editor.cpp]: Don't hesitate to make a PR directly for such things, I can easily forget about a comment I read and answered while busy with other things.

Anyway, I remembered now so I'll fix it. There might be more such wrongly removed bindings, so feel free to send PRs if you stumble on them.

Comment (akien-mga) [reply editor/groups_editor.cpp]: Fixed by #36758. I re-reviewed the list of removed bindings that do not start with an underscore and it seems good now.

There might still be bindings starting with underscore that were removed even though they'd still be used by MessageQueue or UndoRedo.
